ToolBox/CoinApi/CoinApiDataQueueHandler.cs

```diff
@@ -326,5 +341,94 @@ private void OnError(object sender, Exception e)
+        }
+
+        public IEnumerable<BaseData> GetHistory(HistoryRequest historyRequest)
+        {
+            if (historyRequest.Symbol.SecurityType != SecurityType.Crypto)
+            {
+                Log.Error($"CoinApiDataQueueHandler.GetHistory(): Invalid security type {historyRequest.Symbol.SecurityType}");
+                yield break;
+            }
+
+            if (historyRequest.Resolution == Resolution.Tick)
+            {
+                Log.Error("CoinApiDataQueueHandler.GetHistory(): Historical ticks can not be provided, only OHLCV timeseries");
+                yield break;
+            }
+
+            var coinApiSymbol = _symbolMapper.GetBrokerageSymbol(historyRequest.Symbol);
+            var coinApiPeriod = _ResolutionToCoinApiPeriodMappings[historyRequest.Resolution];
+
+            // Time must be in ISO 8601 format
+            var coinApiStartTime = historyRequest.StartTimeUtc.ToStringInvariant("s");
+            var coinApiEndTime = historyRequest.EndTimeUtc.ToStringInvariant("s");
+
+            // Construct URL for rest request
+            var baseUrl =
+                "https://rest.coinapi.io/v1/ohlcv/" +
+                $"{coinApiSymbol}/history?period_id={coinApiPeriod}" +
+                $"&time_start={coinApiStartTime}&time_end={coinApiEndTime}";
+
+            // Execute
+            var client = new RestClient(baseUrl);
+            var restRequest = new RestRequest(Method.GET);
+            restRequest.AddHeader("X-CoinAPI-Key", _apiKey);
+            var response = client.Execute(restRequest);
+
+            // Log the information associated with the API Key's rest call limits.
+            TraceRestUsage(response);
+
+            // Deserialize to array
+            var coinApiHistoryBars = JsonConvert.DeserializeObject<HistoricalDataMessage[]>(response.Content);
+
+            // Can be no historical data for a short period interval
+            if (!coinApiHistoryBars.Any())
+            {
+                Log.Error("CoinApiDataQueueHandler.GetHistory(): API returned no data for the requested period");
+                yield break;
+            }
+
+            foreach (var ohlcv in coinApiHistoryBars)
+            {
+                yield return ohlcv.ToTradeBar();
+            }
+        }
+
+        #endregion
+
+        private void TraceRestUsage(IRestResponse response)
+        {
+            var total = GetHttpHeaderValue(response, "x-ratelimit-limit");
+            var used = GetHttpHeaderValue(response, "x-ratelimit-used");
+            var remaining = GetHttpHeaderValue(response, "x-ratelimit-remaining");
+
+            Log.Trace($"CoinApiDataQueueHandler.TraceRestUsage(): Used {used}, Remaining {remaining}, Total {total}");
+        }
+
+        private string GetHttpHeaderValue(IRestResponse response, string propertyName)
+        {
+            return response.Headers.ToList()
+                .Find(x => x.Name == propertyName)
+                .Value.ToString();
```

Comment:
This will throw null ref if the property name is not present, suggest adding a check for it
```
        [Test]
        public void NullReference()
        {
            var l = new List<Parameter>();
            var tt = l.Find(s => s.Name == "pinocho").Value.ToString();
        }
```